README.md
[![Build Status](https://travis-ci.org/Manuel-777/MTG-Arena-Tool.svg?branch=master)](https://travis-ci.org/Manuel-777/MTG-Arena-Tool)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

# MTG Arena Tool
MTG Arena Tool is a collection browser, a deck tracker and a statistics manager. Explore which decks you played against and what other players are brewing. MTG Arena Tool is all about improving your Magic Arena experience.

![History Screen](/Readme/screenshot_1.png)


## Download
You can find all stable and pre-production releases [right here on GitHub](https://github.com/Manuel-777/MTG-Arena-Tool/releases).
  - [Latest Stable Version](https://github.com/Manuel-777/MTG-Arena-Tool/releases/latest)

Once downloaded the installer should simply install and run immediately. The app will read your user data and warn you if anything goes wrong.

### Linux/Gnome users
After Gnome 3.26, legacy tray icons were deprecated, so users will need to install `libindicator-gtk3` to see the icon again. (packages are available in all managers)

### Source
If you would like to compile and run the latest source, see [the contribution guidelines on how to set up your machine.](./CONTRIBUTING.md#set-up-your-machine)


## Documentation
Visit https://mtgatool.com/ to learn more or join [our community discussion on Discord](https://discord.gg/K9bPkJy).

### Troubleshooting
Please see [our troubleshooting guide.](./TROUBLESHOOTING.md)

### Contributing
Please see [our contributing guide.](./CONTRIBUTING.md)


## License

[GPLv3](./LICENSE.md)

### Disclaimer
Even though no official statement has been made about third party software by MTG Arena developers, I am obliged to put a warning about the use of this software.

It is not stated if it is legal or allowed by Wizards of the Coast to use Deck Trackers and other tools alike while playing MTG Arena, therefore MTG Arena Tool developers are not responsible if your account gets banned, locked or suspended for using this software. `Use at your own risk.`

MtG Arena Tool is unofficial Fan Content permitted under the Fan Content Policy. Not approved/endorsed by Wizards. Portions of the materials used are property of Wizards of the Coast. ©Wizards of the Coast LLC.

Please read about [our Privacy Policy and How we use your data here](./PRIVACY.md)

### Credits
[Electron](https://electronjs.org/)

[Electron Store by Sindre Sorhus](https://github.com/sindresorhus/electron-store)

[Conic Gradient Polyfill by Lea Verou](https://leaverou.github.io/conic-gradient/)

[vanilla-picker by Andreas Borgen, Adam Brooks](https://vanilla-picker.js.org/)

[Draft ranking by Magic Community Set Reviews](https://www.mtgcommunityreview.com/)

[Scryfall.com](http://scryfall.com) in particular, for making an absolutely stunning database of every single card in the multiverse.


## Contact
You can find me at any of these;
[Twitter](https://twitter.com/MEtchegaray7)
[Discord](https://discord.gg/K9bPkJy)
[mtgatool@gmail.com](mailto:mtgatool@gmail.com)


CHANGELOG.md
## <small>2.8.5 (2019-07-26)</small>

* bugfix add new gre message regexes ([e7144fe](https://github.com/Manuel-777/MTG-Arena-Tool/commit/e7144fe))
* bugfix force_open_settings index nums ([07bd6c9](https://github.com/Manuel-777/MTG-Arena-Tool/commit/07bd6c9))
* bugfix handle null notifications messages ([85269bd](https://github.com/Manuel-777/MTG-Arena-Tool/commit/85269bd))
* update PACK_SIZES with m20 ([62d6a51](https://github.com/Manuel-777/MTG-Arena-Tool/commit/62d6a51))



## <small>2.8.4 (2019-07-25)</small>

* add custom date since to filter panel ([053b101](https://github.com/Manuel-777/MTG-Arena-Tool/commit/053b101))
* add pikaday module dependency ([93bd590](https://github.com/Manuel-777/MTG-Arena-Tool/commit/93bd590))
* add pseudo-settings for logUri and log_locale_format ([3e0e7eb](https://github.com/Manuel-777/MTG-Arena-Tool/commit/3e0e7eb))
* background-util.parseWotcTime uses settings.log_locale_format first ([0477eaa](https://github.com/Manuel-777/MTG-Arena-Tool/commit/0477eaa))
* bugfix player_data_refresh background ([2dcf879](https://github.com/Manuel-777/MTG-Arena-Tool/commit/2dcf879))
* bump settings tab index numbers ([95a5df6](https://github.com/Manuel-777/MTG-Arena-Tool/commit/95a5df6))
* fix parse doc link ([a037491](https://github.com/Manuel-777/MTG-Arena-Tool/commit/a037491))
* Friendly format checker ([1a46b8d](https://github.com/Manuel-777/MTG-Arena-Tool/commit/1a46b8d))
* more description ([6162b0b](https://github.com/Manuel-777/MTG-Arena-Tool/commit/6162b0b))
* move last log timestamp and format to player data ([d202c1a](https://github.com/Manuel-777/MTG-Arena-Tool/commit/d202c1a))
* new data settings tab ([e4a4896](https://github.com/Manuel-777/MTG-Arena-Tool/commit/e4a4896))
* open button, some alignment ([578e7bf](https://github.com/Manuel-777/MTG-Arena-Tool/commit/578e7bf))
* renderer-util.renderLogInput ([5e1aaff](https://github.com/Manuel-777/MTG-Arena-Tool/commit/5e1aaff))
* renderer-util.showDatepicker ([ac43143](https://github.com/Manuel-777/MTG-Arena-Tool/commit/ac43143))
* renderer.no_log uses renderer-util.renderLogInput ([fd7cc66](https://github.com/Manuel-777/MTG-Arena-Tool/commit/fd7cc66))
* settings display tweaks ([8ced802](https://github.com/Manuel-777/MTG-Arena-Tool/commit/8ced802))
* settings page uses renderer-util.renderLogInput ([c27fcdc](https://github.com/Manuel-777/MTG-Arena-Tool/commit/c27fcdc))
* settings tab index constants ([734bbee](https://github.com/Manuel-777/MTG-Arena-Tool/commit/734bbee))
* Update package-lock.json ([ddbc940](https://github.com/Manuel-777/MTG-Arena-Tool/commit/ddbc940))



## <small>2.8.3 (2019-07-21)</small>

* [BUGFIX] Fixes action log share icon location reference ([1d47626](https://github.com/Manuel-777/MTG-Arena-Tool/commit/1d47626))
* add custom sort options to decks page ([5ff7fb5](https://github.com/Manuel-777/MTG-Arena-Tool/commit/5ff7fb5))
* add margin for deck crafting costs ([460aa0d](https://github.com/Manuel-777/MTG-Arena-Tool/commit/460aa0d))
* bugfix booster estimate accounts for owned wildcards ([231ce36](https://github.com/Manuel-777/MTG-Arena-Tool/commit/231ce36))
* bugfix db singleton handles filtering out bad sets ([47bab9b](https://github.com/Manuel-777/MTG-Arena-Tool/commit/47bab9b))
* bugfix player_data_refresh background ([4060c6e](https://github.com/Manuel-777/MTG-Arena-Tool/commit/4060c6e))
* bugfix ranked estimates on history page ([6231310](https://github.com/Manuel-777/MTG-Arena-Tool/commit/6231310))
* filter-panel and aggregator support inclusive color matches ([413cebb](https://github.com/Manuel-777/MTG-Arena-Tool/commit/413cebb))
* fix tray icon for osx ([73a5b62](https://github.com/Manuel-777/MTG-Arena-Tool/commit/73a5b62))
* More detailed parse errors ([50685ce](https://github.com/Manuel-777/MTG-Arena-Tool/commit/50685ce))
* typo fix ([2b0a80f](https://github.com/Manuel-777/MTG-Arena-Tool/commit/2b0a80f))
* Update arena-log-decoder.js ([d0b4893](https://github.com/Manuel-777/MTG-Arena-Tool/commit/d0b4893))
* Update database.json ([6a51ed5](https://github.com/Manuel-777/MTG-Arena-Tool/commit/6a51ed5))
* use plus symbol instead of multicolored ([cb4a8c9](https://github.com/Manuel-777/MTG-Arena-Tool/commit/cb4a8c9))



## <small>2.8.2 (2019-07-18)</small>

* add filter for no archetype set ([f7729c3](https://github.com/Manuel-777/MTG-Arena-Tool/commit/f7729c3))
* add mastery logo to economy view ([7e92a5e](https://github.com/Manuel-777/MTG-Arena-Tool/commit/7e92a5e))
* background handles TrackRewardTier.Updated ([d4ae224](https://github.com/Manuel-777/MTG-Arena-Tool/commit/d4ae224))
* bugfix autocomplete should not suggest all or no archetype ([1e11d2e](https://github.com/Manuel-777/MTG-Arena-Tool/commit/1e11d2e))
* bugfix economy page move xp to inventory txns ([07a3f6a](https://github.com/Manuel-777/MTG-Arena-Tool/commit/07a3f6a))
* bugfix mastery track timestamps ([0b00f85](https://github.com/Manuel-777/MTG-Arena-Tool/commit/0b00f85))
* bugfix only default match tags to oppDeck.archetype once ([fc95c52](https://github.com/Manuel-777/MTG-Arena-Tool/commit/fc95c52))
* copy date constants into shared ([47fe739](https://github.com/Manuel-777/MTG-Arena-Tool/commit/47fe739))
* date filter last 24 hours ([afca7ba](https://github.com/Manuel-777/MTG-Arena-Tool/commit/afca7ba))
* economy page consistent formatPercent ([f667dd7](https://github.com/Manuel-777/MTG-Arena-Tool/commit/f667dd7))
* economy page displays TrackRewardTier.Updated ([844d027](https://github.com/Manuel-777/MTG-Arena-Tool/commit/844d027))
* Fix output log dialog ([5492c46](https://github.com/Manuel-777/MTG-Arena-Tool/commit/5492c46)), closes [#484](https://github.com/Manuel-777/MTG-Arena-Tool/issues/484)
* improved context and pretty context for track progress ([779baf7](https://github.com/Manuel-777/MTG-Arena-Tool/commit/779baf7))
* little warning ([f573d7a](https://github.com/Manuel-777/MTG-Arena-Tool/commit/f573d7a))

shared/constants.js
exports.IPC_BACKGROUND = 0;
exports.IPC_MAIN = 1;
exports.IPC_OVERLAY = 2;

// Colour indices
exports.WHITE = 1;
exports.BLUE = 2;
exports.BLACK = 3;
exports.RED = 4;
exports.GREEN = 5;

// Magic constant to represent auth token in form
exports.HIDDEN_PW = "********";

exports.ENUMS = {
  CounterType: {
    "1": "+1/+1",
    "2": "-1/-1",
    "3": "Poison",
    "4": "Wind",
    "5": "Time",
    "6": "Fade",
    "7": "Loyalty",
    "8": "Wish",
    "9": "Age",
    "10": "Aim",
    "11": "Arrow",
    "12": "Arrowhead",
    "13": "Awakening",
    "14": "Blaze",
    "15": "Blood",
    "16": "Bounty",
    "17": "Bribery",
    "18": "Carrion",
    "19": "Charge",
    "20": "Control",
    "21": "Corpse",
    "22": "Credit",
    "23": "Cube",
    "24": "Currency",
    "25": "Death",
    "26": "Delay",
    "27": "Depletion",
    "28": "Despair",
    "29": "Devotion",
    "30": "Divinity",
    "31": "Doom",
    "32": "Dream",
    "33": "Echo",
    "34": "Elixir",
    "35": "Energy",
    "36": "Eon",
    "37": "Eyeball",
    "38": "Fate",
    "39": "Feather",
    "40": "Filibuster",
    "41": "Flame",
    "42": "Flood",
    "43": "Fungus",
    "44": "Fuse",
    "45": "Glyph",
    "46": "Gold",
    "47": "Growth",
    "48": "Hatchling",
    "49": "Healing",
    "50": "Hoofprint",
    "51": "Hourglass",
    "52": "Hunger",
    "53": "Ice",
    "54": "Infection",
    "55": "Intervention",
    "56": "Javelin",
    "57": "Ki",
    "58": "Level",
    "59": "Luck",
    "60": "Magnet",
    "61": "Mannequin",
    "62": "Matrix",
    "63": "May",
    "64": "Mine",
    "65": "Mining",
    "66": "Mire",
    "67": "Muster",
    "68": "Net",
    "69": "Omen",
    "70": "Ore",
    "71": "Page",
    "72": "Pain",
    "73": "Paralyzation",
    "74": "Petal",
    "75": "Petrification",
    "76": "Phylactery",
    "77": "Pin",
    "78": "Plague",
    "79": "Polyp",
    "80": "Pressure",
    "81": "Pupa",
    "82": "Quest",
    "83": "Scream",
    "84": "Scroll",
    "85": "Shell",
    "86": "Shield",
    "87": "Shred",
    "88": "Sleep",
    "89": "Sleight",
    "90": "Slime",
    "91": "Soot",
    "92": "Spell",
    "93": "Spore",
    "94": "Storage",
    "95": "Strife",
    "96": "Study",
    "97": "Theft",
    "98": "Tide",
    "100": "Tower",
    "101": "Training",
    "102": "Trap",
    "103": "Treasure",
    "104": "Verse",
    "105": "Vitality",
    "106": "Wage",
    "107": "Winch",
    "108": "Lore",
    "109": "+1/+2",
    "110": "+0/+1",
    "111": "+0/+2",
    "112": "+1/+0",
    "113": "+2/+2",
    "114": "-0/-1",
    "115": "-0/-2",
    "116": "-1/-0",
    "117": "-2/-1",
    "118": "-2/-2",
    "119": "Manifestation",
    "120": "Gem",
    "121": "Crystal",
    "122": "Isolation",
    "123": "Hour",
    "124": "Unity",
    "125": "Velocity",
    "126": "Brick",
    "127": "Landmark",
    "128": "Prey",
    "129": "Silver",
    "130": "Egg",
    "131": "Hit",
    "132": "PlaceholderCounterType1",
    "133": "PlaceholderCounterType2",
    "134": "PlaceholderCounterType3",
    "135": "PlaceholderCounterType4",
    "136": "PlaceholderCounterType5"
  },
  SuperType: {
    "1": "Basic",
    "2": "Legendary",
    "3": "Ongoing",
    "4": "Snow",
    "5": "World"
  },
  ResultCode: {
    "0": "",
    "1": "Success",
    "2": "Failure",
    "3": "Creature cannot attack",
    "4": "Attacking costs not paid",
    "5": "Creature cannot block",
    "6": "Blocker cannot block attacker",
    "7": "Damage ordering contains omissions or additions",
    "8": "Damage sources contain omissions",
    "9": "Damage sources contain additions",
    "10":
      "Attempt to assign damage such that non-lethal damage is assigned to anything other than the last recipient in the order",
    "12": "Too many targets selected",
    "13": "Not enough targets selected",
    "14": "Selected target is not valid"
  },
  SubType: {
    "1": "Angel",
    "2": "Archer",
    "3": "Archon",
    "4": "Artificer",
    "5": "Assassin",
    "6": "Aura",
    "7": "Basilisk",
    "8": "Bat",
    "9": "Bear",
    "10": "Beast",
    "11": "Berserker",
    "12": "Bird",
    "13": "Boar",
    "14": "Cat",
    "15": "Chandra",
    "16": "Cleric",
    "17": "Construct",
    "18": "Crocodile",
    "19": "Demon",
    "20": "Djinn",
    "21": "Dragon",
    "22": "Drake",
    "23": "Druid",
    "24": "Fish",
    "25": "Elemental",
    "26": "Elephant",
    "27": "Elf",
    "28": "Equipment",
    "29": "Forest",
    "30": "Garruk",
    "31": "Gate",
    "32": "Giant",
    "33": "Gideon",
    "34": "Goblin",
    "35": "Golem",
    "36": "Griffin",
    "37": "Horse",
    "38": "Hound",
    "39": "Human",
    "40": "Hydra",
    "41": "Illusion",
    "42": "Insect",
    "43": "Island",
    "44": "Jace",
    "45": "Knight",
    "46": "Merfolk",
    "47": "Minotaur",
    "48": "Monk",
    "49": "Mountain",
    "50": "Ogre",
    "51": "Ooze",
    "52": "Pegasus",
    "53": "Phoenix",
    "54": "Plains",
    "55": "Rhino",
    "56": "Rogue",
    "57": "Salamander",
    "58": "Scout",
    "59": "Serpent",
    "60": "Shade",
    "61": "Shaman",
    "62": "Siren",
    "63": "Skeleton",
    "64": "Soldier",
    "65": "Sorin",
    "66": "Sphinx",
    "67": "Spider",
    "68": "Spirit",
    "69": "Swamp",
    "70": "Tower",
    "71": "Treefolk",
    "72": "Troll",
    "73": "Urza's",
    "74": "Vampire",
    "75": "Vedalken",
    "76": "Wall",
    "77": "Warrior",
    "78": "Wizard",
    "79": "Wolf",
    "80": "Wurm",
    "81": "Zombie",
    "82": "Mine",
    "83": "Power-Plant",
    "84": "Saproling",
    "85": "Avatar",
    "86": "Sliver",
    "87": "Samurai",
    "88": "Pest",
    "89": "Thalakos",
    "90": "Dauthi",
    "91": "Minion",
    "92": "Advisor",
    "93": "Ajani",
    "94": "Alara",
    "95": "Ally",
    "97": "Antelope",
    "98": "Ape",
    "99": "Arcane",
    "100": "Arkhos",
    "101": "Ashiok",
    "102": "Assembly-Worker",
    "103": "Atog",
    "104": "Aurochs",
    "105": "Azgol",
    "106": "Badger",
    "107": "Barbarian",
    "108": "Beeble",
    "109": "Belenon",
    "110": "Bolas",
    "111": "Bolass",
    "112": "Bringer",
    "113": "Brushwagg",
    "114": "Camel",
    "115": "Carrier",
    "116": "Centaur",
    "117": "Cephalid",
    "118": "Chimera",
    "119": "Cockatrice",
    "120": "Crab",
    "121": "Curse",
    "122": "Cyclops",
    "123": "Desert",
    "124": "Devil",
    "125": "Dominaria",
    "126": "Domri",
    "127": "Dreadnought",
    "128": "Drone",
    "129": "Dryad",
    "130": "Dwarf",
    "131": "Efreet",
    "132": "Elder",
    "133": "Eldrazi",
    "134": "Elk",
    "135": "Elspeth",
    "136": "Equilor",
    "137": "Ergamon",
    "138": "Eye",
    "139": "Fabacin",
    "140": "Faerie",
    "141": "Ferret",
    "142": "Flagbearer",
    "143": "Fortification",
    "144": "Fox",
    "145": "Frog",
    "146": "Fungus",
    "147": "Gargoyle",
    "148": "Gnome",
    "149": "Goat",
    "150": "God",
    "151": "Gorgon",
    "152": "Gremlin",
    "153": "Hag",
    "154": "Harpy",
    "155": "Hellion",
    "156": "Hippo",
    "157": "Hippogriff",
    "158": "Homarid",
    "159": "Homunculus",
    "160": "Horror",
    "161": "Hyena",
    "162": "Imp",
    "163": "Incarnation",
    "164": "Innistrad",
    "165": "Iquatana",
    "166": "Ir",
    "167": "Jellyfish",
    "168": "Juggernaut",
    "169": "Kaldheim",
    "170": "Kamigawa",
    "171": "Karn",
    "172": "Karsus",
    "173": "Kavu",
    "174": "Kephalai",
    "175": "Kirin",
    "176": "Kithkin",
    "177": "Kobold",
    "178": "Kolbahan",
    "179": "Kor",
    "180": "Koth",
    "181": "Kraken",
    "182": "Kyneth",
    "183": "Lair",
    "184": "Lammasu",
    "185": "Leech",
    "186": "Leviathan",
    "187": "Lhurgoyf",
    "188": "Licid",
    "189": "Liliana",
    "190": "Lizard",
    "191": "Locus",
    "192": "Lorwyn",
    "193": "Luvion",
    "194": "Manticore",
    "195": "Masticore",
    "196": "Meditation",
    "197": "Mercadia",
    "198": "Mercenary",
    "199": "Metathran",
    "200": "Mirrodin",
    "201": "Moag",
    "202": "Monger",
    "203": "Mongoose",
    "204": "Mongseng",
    "205": "Moonfolk",
    "206": "Muraganda",
    "207": "Mutant",
    "208": "Myr",
    "209": "Mystic",
    "210": "Nautilus",
    "211": "Nephilim",
    "212": "New",
    "213": "Nightmare",
    "214": "Nightstalker",
    "215": "Ninja",
    "216": "Nissa",
    "217": "Noggle",
    "218": "Nomad",
    "219": "Nymph",
    "220": "Octopus",
    "221": "Orc",
    "222": "Orgg",
    "223": "Ouphe",
    "224": "Ox",
    "225": "Oyster",
    "226": "Phelddagrif",
    "227": "Phyrexia",
    "228": "Pirate",
    "229": "Plant",
    "230": "Praetor",
    "231": "Pyrulea",
    "232": "Rabbit",
    "233": "Rabiah",
    "234": "Ral",
    "235": "Rat",
    "236": "Rath",
    "237": "Ravnica",
    "238": "Realm",
    "239": "Rebel",
    "240": "Regatha",
    "241": "Rigger",
    "242": "Sable",
    "243": "Sarkhan",
    "244": "Satyr",
    "245": "Scarecrow",
    "246": "Scorpion",
    "247": "Segovia",
    "248": "Serras",
    "249": "Shadowmoor",
    "250": "Shandalar",
    "251": "Shapeshifter",
    "252": "Sheep",
    "253": "Shrine",
    "254": "Slith",
    "255": "Slug",
    "256": "Snake",
    "257": "Soltari",
    "258": "Spawn",
    "259": "Specter",
    "260": "Spellshaper",
    "261": "Spike",
    "262": "Sponge",
    "263": "Squid",
    "264": "Squirrel",
    "265": "Starfish",
    "266": "Surrakar",
    "267": "Tamiyo",
    "268": "Tezzeret",
    "269": "Thopter",
    "270": "Thrull",
    "271": "Tibalt",
    "272": "Trap",
    "273": "Turtle",
    "274": "Ulgrotha",
    "275": "Unicorn",
    "276": "Valla",
    "277": "Venser",
    "278": "Viashino",
    "279": "Volver",
    "280": "Vraska",
    "281": "Vryn",
    "282": "Weird",
    "283": "Werewolf",
    "284": "Whale",
    "285": "Wildfire",
    "286": "Wolverine",
    "287": "Wombat",
    "288": "Worm",
    "289": "Wraith",
    "290": "Xenagos",
    "291": "Xerex",
    "292": "Yeti",
    "293": "Zendikar",
    "294": "Zubera",
    "295": "Germ",
    "296": "Contraption",
    "297": "Citizen",
    "298": "Coward",
    "299": "Deserter",
    "300": "Prism",
    "301": "Reflection",
    "302": "Sand",
    "303": "Serf",
    "304": "Dack",
    "305": "Kiora",
    "306": "AllCreatureTypes",
    "307": "Blinkmoth",
    "308": "Camarid",
    "309": "Caribou",
    "310": "Graveborn",
    "311": "Lamia",
    "312": "Orb",
    "313": "Pentavite",
    "314": "Pincher",
    "315": "Splinter",
    "316": "Survivor",
    "317": "Tetravite",
    "318": "Triskelavite",
    "319": "Scion",
    "320": "Processor",
    "321": "Arlinn",
    "322": "Mole",
    "323": "Nahiri",
    "324": "Clue",
    "325": "Teferi",
    "326": "Daretti",
    "327": "Freyalise",
    "328": "Nixilis",
    "329": "Narset",
    "330": "Ugin",
    "331": "Vehicle",
    "332": "Servo",
    "333": "Dovin",
    "334": "Saheeli",
    "335": "Monkey",
    "336": "Aetherborn",
    "337": "Pilot",
    "338": "Jackal",
    "339": "Naga",
    "340": "Cartouche",
    "341": "Samut",
    "342": "Dinosaur",
    "343": "Treasure",
    "344": "Huatli",
    "345": "Angrath",
    "346": "Trilobite",
    "347": "Saga",
    "348": "Jaya",
    "349": "Vivien",
    "350": "Egg",
    "351": "PlaceholderSubType1",
    "352": "PlaceholderSubType2",
    "353": "PlaceholderSubType3",
    "354": "PlaceholderSubType4",
    "355": "PlaceholderSubType5"
  },
  FailureReason: {
    "0": "",
    "1": "Request made with out of date game state.",
    "2": "Player has acted out of turn.",
    "3": "Response does not match the pending request.",
    "4": "Attempted to batch actions that must be performed one at at time.",
    "5":
      "Attempted to perform an action not currently on the list of legal actions.",
    "6":
      "An optional field in the message should have been supplied based on the contents of the required fields, but was not.",
    "7": "Selected an option that is not on the list of legal options.",
    "8": "Message contains a bad enum.",
    "9": "Targeted spell or ability does not have sufficient valid targets.",
    "10": "Target specified for out-of-range target index.",
    "11": "Specified target is not on the legal targets list.",
    "12": "Specified mana does not exist.",
    "13": "Specified option selection is not valid.",
    "14": "The message received was not expected by the GRE.",
    "15":
      "Player input is below a specified minimum or above a specified maximum."
  },
  Phase: {
    "0": "",
    "1": "Beginning",
    "2": "1st Main",
    "3": "Combat",
    "4": "2nd Main",
    "5": "Ending"
  },
  MatchState: {
    "0": "",
    "1": "Game In Progress",
    "2": "Game Complete, Match In Progress",
    "3": "Match Complete",
    "4": "Sideboarding"
  },
  Step: {
    "0": "",
    "1": "Untap",
    "2": "Upkeep",
    "3": "Draw",
    "4": "Begin Combat",
    "5": "Declare Attackers",
    "6": "Declare Blockers",
    "7": "Combat Damage",
    "8": "End Combat",
    "9": "End",
    "10": "Cleanup",
    "11": "First Strike Damage"
  },
  CardType: {
    "1": "Artifact",
    "2": "Creature",
    "3": "Enchantment",
    "4": "Instant",
    "5": "Land",
    "6": "Phenomenon",
    "7": "Plane",
    "8": "Planeswalker",
    "9": "Scheme",
    "10": "Sorcery",
    "11": "Tribal",
    "12": "Vanguard"
  },
  Color: { "1": "White", "2": "Blue", "3": "Black", "4": "Red", "5": "Green" }
};

exports.FORMATS = {
  Standard: "Standard",
  TraditionalStandard: "Traditional Standard",
  Draft: "Draft",
  Sealed: "Sealed",
  Pauper: "Pauper",
  Singleton: "Singleton",
  Cascade: "Cascade",
  Pandemonium: "Pandemonium",
  NoInstants: "No Instants",
  DirectGame: "Direct Game",
  precon: "Preconstructed",
  Ravnica: "Ravnica Block",
  Ixalan: "Ixalan Block",
  GRN: "Ravnica Constructed",
  XLN: "Ixalan Constructed"
};

exports.COLORS_ALL = ["w", "u", "b", "r", "g", "c"];
exports.COLORS_BRIEF = ["w", "u", "b", "r", "g"];
exports.RANKS = ["Bronze", "Silver", "Gold", "Platinum", "Diamond", "Mythic"];

exports.OVERLAY_FULL = 0;
exports.OVERLAY_LEFT = 1;
exports.OVERLAY_ODDS = 2;
exports.OVERLAY_SEEN = 3;
exports.OVERLAY_LOG = 4;
exports.OVERLAY_DRAFT = 5;
exports.OVERLAY_DRAFT_BREW = 6;
exports.OVERLAY_MIXED = 7;
exports.OVERLAY_DRAFT_MODES = [
  exports.OVERLAY_DRAFT,
  exports.OVERLAY_DRAFT_BREW
];

exports.ARENA_MODE_IDLE = 0;
exports.ARENA_MODE_MATCH = 1;
exports.ARENA_MODE_DRAFT = 2;

const DRAFT_RANKS = [];
DRAFT_RANKS[12] = "A+";
DRAFT_RANKS[11] = "A";
DRAFT_RANKS[10] = "A-";
DRAFT_RANKS[9] = "B+";
DRAFT_RANKS[8] = "B";
DRAFT_RANKS[7] = "B-";
DRAFT_RANKS[6] = "C+";
DRAFT_RANKS[5] = "C";
DRAFT_RANKS[4] = "C-";
DRAFT_RANKS[3] = "D+";
DRAFT_RANKS[2] = "D";
DRAFT_RANKS[1] = "D-";
DRAFT_RANKS[0] = "F";
exports.DRAFT_RANKS = DRAFT_RANKS;

exports.CARD_TYPE_CODES = ["cre", "lan", "ins", "sor", "enc", "art", "pla"];
exports.CARD_TYPES = [
  "Creatures",
  "Lands",
  "Instants",
  "Sorceries",
  "Enchantments",
  "Artifacts",
  "Planeswalkers"
];
exports.CARD_RARITIES = ["common", "uncommon", "rare", "mythic"];
exports.MANA_COLORS = [
  "#E7CA8E",
  "#AABEDF",
  "#A18E87",
  "#DD8263",
  "#B7C89E",
  "#E3E3E3"
];

exports.MANA = {
  0: "",
  1: "white",
  2: "blue",
  3: "black",
  4: "red",
  5: "green",
  6: "colorless",
  7: "",
  8: "x"
};

exports.RANKS_SORT = {
  Begginer: 0,
  Bronze: 1,
  Silver: 2,
  Gold: 3,
  Platinum: 4,
  Diamond: 5,
  Mythic: 6
};

exports.PACK_SIZES = {
  "Ravnica Allegiance": 14,
  "Guilds of Ravnica": 14,
  "Core Set 2020": 15
};

exports.DEFAULT_TILE = 67003;

exports.CARD_TILE_ARENA = 0;
exports.CARD_TILE_FLAT = 1;

exports.EASING_DEFAULT = "cubicBezier(0.570, 0.165, 0.210, 0.990)";

exports.MAIN_LOGIN = -2;
exports.MAIN_HOME = -1;
exports.MAIN_DECKS = 0;
exports.MAIN_HISTORY = 1;
exports.MAIN_EVENTS = 2;
exports.MAIN_EXPLORE = 3;
exports.MAIN_ECONOMY = 4;
exports.MAIN_COLLECTION = 5;
exports.MAIN_SETTINGS = 6;
exports.MAIN_UPDATE = 9;

exports.SETTINGS_BEHAVIOUR = 1;
exports.SETTINGS_ARENA_DATA = 2;
exports.SETTINGS_OVERLAY = 3;
exports.SETTINGS_VISUAL = 4;
exports.SETTINGS_PRIVACY = 5;
exports.SETTINGS_ABOUT = 6;
exports.SETTINGS_LOGIN = 7;

// Date constants
exports.DATE_LAST_DAY = "Last 24 Hours";
exports.DATE_LAST_30 = "Last 30 Days";
exports.DATE_SEASON = "Current Season";
exports.DATE_ALL_TIME = "All Time";
